File: chela/e2e.py
"""End-to-end crypto for the collaborative terminal share (auth-plane CHUNK 2).

The relay is a dumb public forwarder; it must never see plaintext. This module is
the SINGLE SOURCE OF TRUTH for the on-the-wire crypto contract — every constant
and byte layout here is mirrored exactly by the browser SPA's WebCrypto code
(chela/collab-relay/public/index.html) and locked by the Python↔JS interop vector
suite (tests/test_e2e_vectors.py). Change a byte here → regenerate vectors → update
the SPA, or the two sides silently stop agreeing.

Grounded in docs/AUTH_PLANE_SCOPE.md §5 (Crypto). We roll NOTHING custom: HKDF and
AES-GCM come from `cryptography` (Python) and WebCrypto SubtleCrypto (browser),
both vetted, natively interoperable.

CONTRACT
--------
Pairing secret: 16 random bytes (owner mints via `mint_secret`), shown to the
  joiner as base32 (`pairing_code`) and pasted back (`secret_from_code`).

Key derivation: HKDF-SHA256 with a fixed salt and a room-bound info string yields
  TWO independent AES-256 keys, one per direction, so the two senders can never
  collide on a nonce:
      k_h2j = HKDF(secret, salt=SALT, info="chela-v1|<room>|h2j", L=32)   host→joiner
      k_j2h = HKDF(secret, salt=SALT, info="chela-v1|<room>|j2h", L=32)   joiner→host

Nonce (96-bit, deterministic — never random, never reused for a key): 4 reserved
  zero bytes ∥ 8-byte little-endian per-key monotonic `seq`. A new share = new
  secret = new keys, so the counters reset safely.

Envelope (binary, little-endian): ver(1) ∥ type(1) ∥ seq(8) ∥ ciphertext, where
  ciphertext is AES-256-GCM output with the 128-bit tag appended (the library
  default). GCM AAD = ver(1) ∥ type(1) ∥ seq(8) ∥ utf8(room) — so header
  tampering, cross-room replay, and type confusion ALL fail the tag.

Receiver enforces strictly-increasing `seq` per direction: replays and reorders
  are dropped (gaps are allowed — a late joiner just starts from the current seq).
  A wrong pairing code derives wrong keys → the first frame fails the GCM tag →
  we raise AuthError ("wrong code"), never emit garbage.
"""

from __future__ import annotations

import base64
import re
import secrets

from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

# --- wire constants (mirrored byte-for-byte in the SPA) -----------------------
VER = 1
# Frame types. Non-zero so a zero-initialised buffer never looks like a valid type.
T_OUTPUT = 1     # terminal output bytes  (host→joiner)
T_INPUT = 2      # terminal input bytes   (joiner→host; P1.5 write-grant)
T_META = 3       # JSON grid/title        (host→joiner)
T_PRESENCE = 4   # JSON presence          (either; P2)
T_CTL = 5        # JSON control, e.g. hello (joiner→host)

SALT = b"chela-collab-e2e-v1"   # fixed HKDF salt, identical both sides
KEY_LEN = 32                    # AES-256
NONCE_LEN = 12                  # 96-bit GCM nonce
_RESERVED = b"\x00\x00\x00\x00"  # 4 reserved nonce bytes
HEADER_LEN = 10                 # ver(1)+type(1)+seq(8)
SECRET_LEN = 16                 # pairing secret bytes


class E2EError(Exception):
    """Base for all envelope-processing failures."""


class AuthError(E2EError):
    """GCM tag verification failed — wrong key (wrong pairing code) or tampering."""


class ReplayError(E2EError):
    """seq did not strictly increase — a replay or reorder; dropped."""


class BadFrame(E2EError):
    """Malformed envelope (too short / unknown version)."""


# --- pairing code (base32) ----------------------------------------------------
def mint_secret() -> bytes:
    """A fresh 16-byte pairing secret (owner-side, per share)."""
    return secrets.token_bytes(SECRET_LEN)


def pairing_code(secret: bytes) -> str:
    """Display form: RFC4648 base32, uppercase, padding stripped (16 bytes → 26
    chars). The SPA's base32 decoder mirrors this exactly."""
    return base64.b32encode(secret).decode("ascii").rstrip("=")


def secret_from_code(code: str) -> bytes:
    """Parse a pasted pairing code back to the 16-byte secret. Tolerant of
    whitespace, case, and missing padding (re-padded to a multiple of 8)."""
    s = re.sub(r"\s+", "", code).upper().rstrip("=")
    s += "=" * ((-len(s)) % 8)
    try:
        secret = base64.b32decode(s)
    except Exception as e:  # noqa: BLE001 - normalise to our error type
        raise BadFrame(f"invalid pairing code: {e}") from e
    if len(secret) != SECRET_LEN:
        raise BadFrame(f"pairing code decodes to {len(secret)} bytes, expected {SECRET_LEN}")
    return secret


# --- key derivation -----------------------------------------------------------
def _hkdf(secret: bytes, info: bytes) -> bytes:
    return HKDF(algorithm=SHA256(), length=KEY_LEN, salt=SALT, info=info).derive(secret)


def derive_keys(secret: bytes, room: str) -> tuple[bytes, bytes]:
    """(k_h2j, k_j2h) for a room. info is room-bound so a secret leaked from one
    room can't decrypt another."""
    return (
        _hkdf(secret, f"chela-v1|{room}|h2j".encode()),
        _hkdf(secret, f"chela-v1|{room}|j2h".encode()),
    )


def _nonce(seq: int) -> bytes:
    return _RESERVED + seq.to_bytes(8, "little")


# --- session (one per peer): send on one key, receive on the other ------------
class Session:
    """Seals/opens envelopes for one endpoint. `role` picks which direction key is
    used for sending vs receiving:
        host   → send h2j, recv j2h
        joiner → send j2h, recv h2j
    Not thread-safe for concurrent seal() (the send seq is mutable state); the
    caller serialises seal+send so wire order matches seq order (the receiver
    rejects out-of-order seqs)."""

    def __init__(self, secret: bytes, room: str, *, role: str) -> None:
        if role not in ("host", "joiner"):
            raise ValueError("role must be 'host' or 'joiner'")
        k_h2j, k_j2h = derive_keys(secret, room)
        send_key, recv_key = (k_h2j, k_j2h) if role == "host" else (k_j2h, k_h2j)
        self.room = room
        self._room_aad = room.encode("utf-8")
        self._send = AESGCM(send_key)
        self._recv = AESGCM(recv_key)
        self._send_seq = 0
        self._recv_last = -1   # strictly-increasing gate; first accepted seq is >= 0

    def seal(self, typ: int, plaintext: bytes) -> bytes:
        seq = self._send_seq
        self._send_seq += 1
        header = bytes([VER, typ]) + seq.to_bytes(8, "little")
        ct = self._send.encrypt(_nonce(seq), plaintext, header + self._room_aad)
        return header + ct

    def open(self, envelope: bytes) -> tuple[int, bytes]:
        if len(envelope) < HEADER_LEN:
            raise BadFrame("envelope shorter than header")
        if envelope[0] != VER:
            raise BadFrame(f"unknown version {envelope[0]}")
        typ = envelope[1]
        seq = int.from_bytes(envelope[2:HEADER_LEN], "little")
        # Reject replays/reorders BEFORE the (costly) AEAD open. seq is not yet
        # authenticated, but a forged seq can't advance our window: if the tag
        # fails we raise without touching _recv_last, and a genuine replay (seq
        # <= last) is dropped here cheaply.
        if seq <= self._recv_last:
            raise ReplayError(f"seq {seq} <= last {self._recv_last}")
        header = envelope[:HEADER_LEN]
        try:
            pt = self._recv.decrypt(_nonce(seq), envelope[HEADER_LEN:], header + self._room_aad)
        except InvalidTag as e:
            raise AuthError("GCM tag failed — wrong pairing code or tampered frame") from e
        self._recv_last = seq
        return typ, pt

File: chela/collab-relay/public/e2e.js
// End-to-end crypto for the collaborative terminal share (auth-plane CHUNK 2) —
// the browser/WebCrypto mirror of chela/e2e.py. This is the EXACT module the
// joiner SPA runs AND the module the Python↔JS interop vector suite imports under
// Node (globalThis.crypto.subtle exists in both), so there is one implementation,
// not two that can drift. Every constant and byte layout matches e2e.py; the
// vectors in tests/test_e2e_vectors.py lock the agreement.
//
// Contract (see e2e.py for the authoritative prose):
//   HKDF-SHA256(secret, salt=SALT, info="chela-v1|<room>|{h2j,j2h}") → two AES-256 keys
//   nonce  = 4 reserved zero bytes ∥ 8-byte little-endian seq
//   envelope = ver(1) ∥ type(1) ∥ seq(8 LE) ∥ ciphertext(with 128-bit GCM tag)
//   AAD    = ver(1) ∥ type(1) ∥ seq(8 LE) ∥ utf8(room)
//   strictly-increasing seq per direction; a wrong code fails the first GCM tag.

export const VER = 1;
export const T_OUTPUT = 1, T_INPUT = 2, T_META = 3, T_PRESENCE = 4, T_CTL = 5;
const SALT = new TextEncoder().encode('chela-collab-e2e-v1');
const KEY_LEN = 32, HEADER_LEN = 10, SECRET_LEN = 16;
const te = new TextEncoder();

export class E2EError extends Error {}
export class AuthError extends E2EError {}   // GCM tag failed — wrong code / tamper
export class ReplayError extends E2EError {} // seq did not strictly increase
export class BadFrame extends E2EError {}    // malformed envelope

// --- base32 (RFC4648), mirroring Python base64.b32{encode,decode} -------------
const B32 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
export function b32encode(bytes) {
  let bits = 0, val = 0, out = '';
  for (const b of bytes) {
    val = (val << 8) | b; bits += 8;
    while (bits >= 5) { out += B32[(val >>> (bits - 5)) & 31]; bits -= 5; }
  }
  if (bits > 0) out += B32[(val << (5 - bits)) & 31];
  return out; // no padding (matches pairing_code's rstrip('='))
}
export function secretFromCode(code) {
  const s = code.replace(/\s+/g, '').toUpperCase().replace(/=+$/, '');
  let bits = 0, val = 0; const out = [];
  for (const ch of s) {
    const idx = B32.indexOf(ch);
    if (idx < 0) throw new BadFrame('invalid base32 char in pairing code');
    val = (val << 5) | idx; bits += 5;
    if (bits >= 8) { out.push((val >>> (bits - 8)) & 0xff); bits -= 8; }
  }
  const secret = new Uint8Array(out);
  if (secret.length !== SECRET_LEN) throw new BadFrame(`pairing code decodes to ${secret.length} bytes, expected ${SECRET_LEN}`);
  return secret;
}

// --- key derivation -----------------------------------------------------------
async function hkdfKey(secret, info) {
  const base = await crypto.subtle.importKey('raw', secret, 'HKDF', false, ['deriveBits']);
  const bits = await crypto.subtle.deriveBits(
    { name: 'HKDF', hash: 'SHA-256', salt: SALT, info: te.encode(info) }, base, KEY_LEN * 8);
  return crypto.subtle.importKey('raw', bits, { name: 'AES-GCM' }, false, ['encrypt', 'decrypt']);
}
export async function deriveKeys(secret, room) {
  return {
    h2j: await hkdfKey(secret, `chela-v1|${room}|h2j`),
    j2h: await hkdfKey(secret, `chela-v1|${room}|j2h`),
  };
}

// Diagnostic: the raw HKDF output as hex, for the interop vector suite to assert
// byte-identical key derivation against Python. Not used by the SPA (its keys are
// non-extractable). Same salt/info as hkdfKey — this exercises the real KDF path.
export async function keyHex(secret, room) {
  const hex = async (info) => {
    const base = await crypto.subtle.importKey('raw', secret, 'HKDF', false, ['deriveBits']);
    const bits = await crypto.subtle.deriveBits(
      { name: 'HKDF', hash: 'SHA-256', salt: SALT, info: te.encode(info) }, base, KEY_LEN * 8);
    return [...new Uint8Array(bits)].map((b) => b.toString(16).padStart(2, '0')).join('');
  };
  return { h2j: await hex(`chela-v1|${room}|h2j`), j2h: await hex(`chela-v1|${room}|j2h`) };
}

// seq is a BigInt so the 64-bit counter is exact regardless of magnitude.
function seqBytes(seq) {
  const b = new Uint8Array(8);
  let v = BigInt(seq);
  for (let i = 0; i < 8; i++) { b[i] = Number(v & 0xffn); v >>= 8n; } // little-endian
  return b;
}
function readSeq(bytes) { // bytes[2..10] LE → BigInt
  let v = 0n;
  for (let i = 7; i >= 0; i--) v = (v << 8n) | BigInt(bytes[2 + i]);
  return v;
}
function nonce(seq) { const n = new Uint8Array(12); n.set(seqBytes(seq), 4); return n; } // 4 reserved zeros ∥ seq
function concat(...arrs) {
  const len = arrs.reduce((a, x) => a + x.length, 0);
  const out = new Uint8Array(len); let o = 0;
  for (const x of arrs) { out.set(x, o); o += x.length; }
  return out;
}

// --- session ------------------------------------------------------------------
export class Session {
  // async because WebCrypto key import is async: `await Session.create(...)`.
  static async create(secret, room, role) {
    if (role !== 'host' && role !== 'joiner') throw new Error("role must be 'host' or 'joiner'");
    const keys = await deriveKeys(secret, room);
    const s = new Session();
    s.room = room;
    s._roomAad = te.encode(room);
    s._sendKey = role === 'host' ? keys.h2j : keys.j2h;
    s._recvKey = role === 'host' ? keys.j2h : keys.h2j;
    s._sendSeq = 0n;
    s._recvLast = -1n;
    return s;
  }

  async seal(typ, plaintext) {
    const seq = this._sendSeq; this._sendSeq += 1n;
    const header = concat(new Uint8Array([VER, typ]), seqBytes(seq));
    const aad = concat(header, this._roomAad);
    const ct = new Uint8Array(await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: nonce(seq), additionalData: aad, tagLength: 128 }, this._sendKey, plaintext));
    return concat(header, ct);
  }

  async open(envelope) {
    if (envelope.length < HEADER_LEN) throw new BadFrame('envelope shorter than header');
    if (envelope[0] !== VER) throw new BadFrame(`unknown version ${envelope[0]}`);
    const typ = envelope[1];
    const seq = readSeq(envelope);
    if (seq <= this._recvLast) throw new ReplayError(`seq ${seq} <= last ${this._recvLast}`);
    const header = envelope.subarray(0, HEADER_LEN);
    const aad = concat(header, this._roomAad);
    let pt;
    try {
      pt = await crypto.subtle.decrypt(
        { name: 'AES-GCM', iv: nonce(seq), additionalData: aad, tagLength: 128 },
        this._recvKey, envelope.subarray(HEADER_LEN));
    } catch (_) {
      throw new AuthError('GCM tag failed — wrong pairing code or tampered frame');
    }
    this._recvLast = seq;
    return [typ, new Uint8Array(pt)];
  }
}
